client/client-multi/client-objective-c/Kaa/log/sqlite/SQLiteLogStorage.m

```diff
@@ -0,0 +1,435 @@
+        return logBucket;
+    }
+}
+
+- (void)removeBucketWithId:(int32_t)bucketId {
+    @synchronized(self) {
+        sqlite3_stmt *statement = [self statementForQuery:KAA_DELETE_BY_BUCKET_ID];
+        
+        sqlite3_bind_int64(statement, 1, bucketId);
+        
+        if (sqlite3_step(statement) != SQLITE_DONE) {
+            DDLogError(@"%@ Unable to remove bucket with id [%i]", TAG, bucketId);
+        }
+        
+        int removedRecordsCount = sqlite3_changes(self.database);
+        
+        if (removedRecordsCount > 0) {
+            self.totalRecordCount -= removedRecordsCount;
+            DDLogDebug(@"%@ Removed %i records from storage. Total log record count: %lld",
+                       TAG, removedRecordsCount, self.totalRecordCount);
+        } else {
+            DDLogDebug(@"%@ No records were removed from storage", TAG);
+        }
+        
+        sqlite3_finalize(statement);
+    }
+}
+
+- (void)rollbackBucketWithId:(int32_t)bucketId {
+    @synchronized(self) {
+        DDLogDebug(@"%@ Rollback bucket with id [%i]", TAG, bucketId);
+        
+        sqlite3_stmt *statement = [self statementForQuery:KAA_RESET_BY_BUCKET_ID];
+        
+        sqlite3_bind_int64(statement, 1, bucketId);
+        
+        if (sqlite3_step(statement) != SQLITE_DONE) {
+            DDLogError(@"%@ Unable to rollback bucket with id [%i]", TAG, bucketId);
+        }
+        
+        int resetRecordsCount = sqlite3_changes(self.database);
+        
+        if (resetRecordsCount > 0) {
+            DDLogDebug(@"%@ Reset %i records for bucket with id [%i]", TAG, resetRecordsCount, bucketId);
+            
+            int64_t previouslyConsumedSize = [self.consumedMemoryMap[@(bucketId)] longLongValue];
+            [self.consumedMemoryMap removeObjectForKey:@(bucketId)];
+            
+            self.unmarkedConsumedSize += previouslyConsumedSize;
+            self.unmarkedRecordCount += resetRecordsCount;
+            
+        } else {
+            DDLogDebug(@"%@ No records were reset for bucket with id [%i]", TAG, bucketId);
+        }
+        
+        sqlite3_finalize(statement);
+    }
+}
+
+- (id<LogStorageStatus>)getStatus {
+    return self;
+}
+
+- (int64_t)getConsumedVolume {
+    return _unmarkedConsumedSize;
+}
+
+- (int64_t)getRecordCount {
+    return _unmarkedRecordCount;
+}
+
+- (void)close {
+    if (self.database == NULL) {
+        return;
+    }
+    
+    int result = sqlite3_close(self.database);
+    if (result != SQLITE_OK) {
+        DDLogWarn(@"%@ Failed to close database", TAG);
+    }
+}
+
+- (void)openDBAtPath:(NSString *)path {
+    BOOL dbOpenResult = sqlite3_open([path UTF8String], &_database) == SQLITE_OK;
+    
+    if (!dbOpenResult) {
+        DDLogError(@"%@ Failed to open database at path: %@", TAG, path);
+        return;
+    }
+    
+    [self executeQuery:KAA_CREATE_LOG_TABLE];
+    [self executeQuery:KAA_CREATE_INFO_TABLE];
+    [self executeQuery:KAA_CREATE_BUCKET_ID_INDEX];
```

Comment:
If I understand correctly, it will create table every time database is opened. Is it correct or am I missing something?